## Задание E.
## Гайнутдинова Диана, группа 144.

В данном задании предлагается выполнить статистический анализ двух текстов на русском языке:
    
    
    1) текст главы 2 из книги Льюиса Кэрролла «Алиса в Стране Чудес»
    2) текст по выбору.

В первую очередь будем рассматривать текст из книги Льюиса Кэррола. Он прикреплен к отчету в текстовом файле под названием alice.txt.
В задании было предложено выбрать морфопроцессор. Здесь будет использоваться pymorphy, написанном на Python.

In [272]:
import pymorphy2
import string
import pandas as pd
morph = pymorphy2.MorphAnalyzer()


    
Для начала нам важно выделить все словоформы данного текста. Если разделить текст по пробелам, то в нашем списке словоформы будут 
выглядеть не совсем правильно. Например, слова, после которых стоит знак препинания будут записаны вместе с ним. Нужно обработать наш 
список словоформ так, чтобы с ним можно было работать в дальнейшем. Ниже приведены этапы обработки. 

In [273]:
alice_text = open("/Users/dianagajnutdinova/alice.txt").read()
words = alice_text.split()

В ячейке ниже мы получаем список разделенных слов. Теперь нам нужно провести анализ всех словоформ. Сделаем таблицу из нашего анализа. 
Однако стоит заметить, что в списке много повторений, например одинаковые предлоги встречаются очень много раз. Поэтому нам стоит
сделать уникальное множество словоформ.

In [274]:
exclude = set(string.punctuation)
exclude.add('- ')
exclude.remove('-')
def test_set(s):
    return ''.join(ch for ch in s if ch not in exclude)
words = [test_set(i) for i in words]
words = [i for i in words if i != '' ]

In [275]:
print (len(words), 'длина неуникального списка')
forms = set(words)
print (len(forms), 'длина уникального списка')


2216 длина неуникального списка
1048 длина уникального списка


Длина списка сократилась практически в два раза после того, как мы убрали повторения. Давайте теперь займемся созданием таблицы с анализом всех словоформ текста. 

In [276]:
forms = list(forms)
bias = pd.DataFrame(columns=['Слово','Часть речи','Одушевленность', 'Вид','Падеж','Род','Наклонение','Число',
                             'Лицо','Время','Переходность','Залог','Нормальная форма'],
                      index=[i for i in range(len(forms))])

In [277]:
bias["Слово"] = forms
bias["Часть речи"] = [morph.parse(i)[0].tag.POS for i in forms]
bias['Одушевленность'] = [morph.parse(i)[0].tag.animacy for i in forms]
bias['Вид'] = [morph.parse(i)[0].tag.aspect for i in forms]
bias['Падеж'] = [morph.parse(i)[0].tag.case for i in forms]
bias['Род'] = [morph.parse(i)[0].tag.gender for i in forms]
bias['Наклонение'] = [morph.parse(i)[0].tag.mood for i in forms]
bias['Число'] = [morph.parse(i)[0].tag.number for i in forms]
bias['Лицо'] = [morph.parse(i)[0].tag.person for i in forms]
bias['Время'] = [morph.parse(i)[0].tag.tense for i in forms]
bias['Переходность'] = [morph.parse(i)[0].tag.transitivity for i in forms]
bias['Залог'] = [morph.parse(i)[0].tag.voice for i in forms]
bias["Нормальная форма"] = [morph.parse(i)[0].normal_form for i in forms]

In [278]:
bias.to_html("alice.html")

Таблица оформлена. Для удобства ее лучше посмотреть в отдельном файле, который называется alice.html.
Каждое словоформа содержит 12 признаков, такие как часть речи, вид, время, залог и т. д.


Теперь соберем статистические характеристики текста. Всего будут подсчитаны 9 характеристик. 


1) Длина предложения.


2) Наиболее частый падеж.


3) Относительная частота падежей.


4) Процент разных частей речи.


5) Самые частотные слова и их относительная частота.


6) Самые частотные слова основных частей речи (существительные, прилагательные, наречия, глаголы).


7) Общее число словоупотреблений. (число слов в тексте)


8) Число различных словоформ.


9) Наиболее частотные морфологические формы глаголов (время/лицо/число).

Для того, чтобы посчитать все интересные нам характеристики, нужно немного исправить таблицу.


In [279]:
new_bias = pd.DataFrame(columns=['Слово','Часть речи','Одушевленность', 'Вид','Падеж','Род','Наклонение','Число',
                             'Лицо','Время','Переходность','Залог','Нормальная форма'],
                      index=[i for i in range(len(words))])
new_bias["Слово"] = words
new_bias["Часть речи"] = [morph.parse(i)[0].tag.POS for i in words]
new_bias['Одушевленность'] = [morph.parse(i)[0].tag.animacy for i in words]
new_bias['Вид'] = [morph.parse(i)[0].tag.aspect for i in words]
new_bias['Падеж'] = [morph.parse(i)[0].tag.case for i in words]
new_bias['Род'] = [morph.parse(i)[0].tag.gender for i in words]
new_bias['Наклонение'] = [morph.parse(i)[0].tag.mood for i in words]
new_bias['Число'] = [morph.parse(i)[0].tag.number for i in words]
new_bias['Лицо'] = [morph.parse(i)[0].tag.person for i in words]
new_bias['Время'] = [morph.parse(i)[0].tag.tense for i in words]
new_bias['Переходность'] = [morph.parse(i)[0].tag.transitivity for i in words]
new_bias['Залог'] = [morph.parse(i)[0].tag.voice for i in words]
new_bias["Нормальная форма"] = [morph.parse(i)[0].normal_form for i in words]

### 1) Длина предложения.

In [280]:
len(alice_text.replace('\n   ','').split('.'))


146

 Выше мы высчитываем количество предложений в тексте - 146. Посчитаем длину каждого предложения, а потом вычислим среднюю длину.


In [281]:
lens = [len(i.split(' ')) for i in alice_text.replace('\n   ','').split('.')]
c = 0
for i in lens:
    c += i
c/len(alice_text.replace('\n   ','').split('.'))

18.04794520547945

Получаем, что средняя длина предложений - 18 слов. 

### 2) Наиболее частый падеж.

In [282]:
new_bias["Падеж"].describe()

count      787
unique       8
top       nomn
freq       385
Name: Падеж, dtype: object

Nominative case(именительный падеж) - самый частый падеж среди слов данного текста. Он встречается 385 раз из 787 слов, 
у которых анализатор смог определить падеж. Это состоявляет практически 48%.

### 3) Относительная частота падежей.

In [283]:

cases_table = pd.DataFrame(columns=['Частота'], index = ['Именительный падеж',
'Родительный падеж','Винительный падеж','Дательный падеж','Творительный падеж','Предложный падеж'])
freq = [len(new_bias[new_bias["Падеж"]=='nomn'])/787, len(new_bias[new_bias["Падеж"]=='gent'])/787,len(new_bias[new_bias["Падеж"]=='accs'])/787,
       len(new_bias[new_bias["Падеж"]=='datv'])/787,len(new_bias[new_bias["Падеж"]=='ablt'])/787,len(new_bias[new_bias["Падеж"]=='loct'])/787]
cases_table['Частота'] = freq
cases_table

Частота
Именительный падеж  0.489199
Родительный падеж   0.177891
Винительный падеж   0.139771
Дательный падеж     0.069886
Творительный падеж  0.058450
Предложный падеж    0.062262

Выше можно увидеть табличку, в которой описаны частоты встречаемых падежей. Наиболее редкий падеж - Предложный падеж.

### 4) Процент разных частей речи.

In [284]:
speech = ['Глагол','Существительное','Прилагательное (полное)','Местоимение-существительное','Причастие (полное)','Предлог',
 'Деепричастие','Наречие','Глагол (инфинитив)','Частица','Прилагательное (краткое)','Союз','Причастие (краткое)','Числительное',
 'Предикатив' ,'Компаратив' ,'Междометие']
speech_part_table = pd.DataFrame(columns=['Процентное содержание'], index = speech)
en_speech = ['VERB', 'NOUN', 'ADJF', 'NPRO', 'PRTF', 'PREP',  'GRND',
       'ADVB', 'INFN', 'PRCL', 'ADJS', 'CONJ', 'PRTS', 'NUMR', 'PRED',
       'COMP', 'INTJ']
freq = [len(new_bias[new_bias["Часть речи"]==i])/787*100 for i in en_speech]
speech_part_table['Процентное содержание'] = freq
speech_part_table

Процентное содержание
Глагол                                   41.296061
Существительное                          46.378653
Прилагательное (полное)                  24.523507
Местоимение-существительное              27.445997
Причастие (полное)                        0.381194
Предлог                                  20.076239
Деепричастие                              1.270648
Наречие                                  26.175349
Глагол (инфинитив)                        8.259212
Частица                                  23.634053
Прилагательное (краткое)                  1.905972
Союз                                     36.340534
Причастие (краткое)                       0.508259
Числительное                              1.270648
Предикатив                                0.889454
Компаратив                                2.160102
Междометие                                2.287166

Наиболее часто встречаемая часть речи - существительное, состовляющее почти 50 процентов. Наименее часто встречаемая - Причастие (меньше одного процента.

### 5) Самые частотные слова и их относительная частота.

In [285]:
from collections import Counter
exclude = set(string.punctuation)
def test_set(s):
    return ''.join(ch for ch in s if ch not in exclude)
words = alice_text.split()
words = [test_set(i) for i in words]
words = [i for i in words if i != '']
cap_words = [word.upper() for word in words] #capitalizes all the words

word_counts = Counter(cap_words) 

In [286]:
for i in range(10):
    print ((word_counts.most_common()[i]), word_counts.most_common()[i][1]/len(words))

('И', 97) 0.04547585560243788
('НЕ', 67) 0.031411157993436474
('ОНА', 59) 0.027660571964369433
('В', 47) 0.02203469292076887
('Я', 44) 0.02062822315986873
('ЧТО', 38) 0.017815283638068447
('АЛИСА', 37) 0.017346460384435068
('А', 31) 0.014533520862634786
('ТАК', 25) 0.011720581340834505
('ЭТО', 24) 0.011251758087201125


Выше представлены наиболее частые слова, число их встречаний в тексте, а также частота их появлений.
Как видно из списка, наиболее частое слово - предлог 'и', встречающийся 97 раза. Одако это состовляет всего лишь 0.04 процента.
Логично, что предлоги в тексте встречаются чаще всех остальных слов.

### 6) Самые частотные слова основных частей речи (существительные, прилагательные, наречия, глаголы).

In [287]:
nouns = new_bias[new_bias["Часть речи"]=='NOUN']
nouns.describe()

Слово Часть речи Одушевленность  Вид Падеж   Род  Наклонение Число  \
count     365        365            365  0.0   365   364         0.0   365   
unique    265          1              2  0.0     8     3         0.0     2   
top     Алиса       NOUN           inan  NaN  nomn  femn         NaN  sing   
freq       28        365            246  NaN   139   181         NaN   317   

        Лицо  Время  Переходность  Залог Нормальная форма  
count    0.0    0.0           0.0    0.0              365  
unique   0.0    0.0           0.0    0.0              214  
top      NaN    NaN           NaN    NaN            алиса  
freq     NaN    NaN           NaN    NaN               29

Самое частотное слово среди существительных - Алиса(встречается 28 раз). :))

In [288]:
adj = new_bias[new_bias["Часть речи"]=='ADJF']
adj.describe()

Слово Часть речи Одушевленность  Вид Падеж   Род  Наклонение Число  \
count    193        193              6  0.0   193   128         0.0   193   
unique   144          1              1  0.0     6     3         0.0     2   
top      все       ADJF           inan  NaN  nomn  femn         NaN  sing   
freq      11        193              6  NaN   101    71         NaN   128   

        Лицо  Время  Переходность  Залог Нормальная форма  
count    0.0    0.0           0.0    0.0              193  
unique   0.0    0.0           0.0    0.0               93  
top      NaN    NaN           NaN    NaN             весь  
freq     NaN    NaN           NaN    NaN               24

Самое частотное слово среди прилагательным анализатор посчитал слово все:), которое встречается 11 раз. 

In [289]:
adv = new_bias[new_bias["Часть речи"]=='ADVB']
adv.describe()

Слово Часть речи  Одушевленность  Вид  Падеж  Род  Наклонение  Число  \
count    206        206             0.0  0.0    0.0  0.0         0.0    0.0   
unique   110          1             0.0  0.0    0.0  0.0         0.0    0.0   
top      тут       ADVB             NaN  NaN    NaN  NaN         NaN    NaN   
freq       9        206             NaN  NaN    NaN  NaN         NaN    NaN   

        Лицо  Время  Переходность  Залог Нормальная форма  
count    0.0    0.0           0.0    0.0              206  
unique   0.0    0.0           0.0    0.0              102  
top      NaN    NaN           NaN    NaN         наверное  
freq     NaN    NaN           NaN    NaN               11

Самое частое наречие - тут, встречается 9 раз.

In [290]:
ver = new_bias[new_bias["Часть речи"]=='VERB']
ver.describe()

Слово Часть речи  Одушевленность   Вид  Падеж   Род Наклонение Число  \
count    325        325             0.0   325    0.0   184        325   325   
unique   231          1             0.0     2    0.0     3          2     2   
top     было       VERB             NaN  perf    NaN  femn       indc  sing   
freq      12        325             NaN   178    NaN   136        309   265   

        Лицо Время Переходность  Залог Нормальная форма  
count    107   309          325    0.0              325  
unique     3     3            2    0.0              193  
top     3per  past         intr    NaN             быть  
freq      53   200          219    NaN               34

Самый частый глагол - было(12 раз).

### 7)Общее число словоупотреблений. (число слов в тексте)

In [291]:
words = alice_text.split()
exclude = set(string.punctuation)
exclude.add('- ')
def test_set(s):
    return ''.join(ch for ch in s if ch not in exclude)
words = [test_set(i) for i in words]
words = [i for i in words if i != '']
print ('Количество словоупотреблений',len(words))


Количество словоупотреблений 2133


### 8) Число различных словоформ.

In [292]:
print ('Количество словоформ',len(forms))


Количество словоформ 1048


### 9) Наиболее частотные морфологические формы глаголов (время/лицо/число).

Опишем поподробней глаголы.

In [293]:
new_bias["Время"].describe()

count      333
unique       3
top       past
freq       212
Name: Время, dtype: object

In [294]:
new_bias["Лицо"].describe()

count      291
unique       3
top       3per
freq       153
Name: Лицо, dtype: object

In [295]:
new_bias["Число"].describe()

count     1121
unique       2
top       sing
freq       916
Name: Число, dtype: object

Самое частое время - прошедшее.


Самое частое лицо - 3-е лицо.


Самое частое число - единственное.


Мы развершили анализ одного из двух тектов. Теперь давайте взглянем на первую главу книги Джека Лондона "Мартин Иден". Она является немного 
больше текста из предыдущего пример. С ним мы, надеюсь, сможем провести некоторый сравнительный анализ между двумя текстами.

In [296]:
martin_text = open("/Users/dianagajnutdinova/text.txt").read()
words = martin_text.split()

In [297]:
exclude = set(string.punctuation)
exclude.add('- ')
exclude.remove('-')
def test_set(s):
    return ''.join(ch for ch in s if ch not in exclude)
words = [test_set(i) for i in words]
words = [i for i in words if i != '' ]
print (len(words), 'длина неуникального списка')
forms = set(words)
print (len(forms), 'длина уникального списка')


3669 длина неуникального списка
1774 длина уникального списка


Как я уже говорила, текст Джека Лондона немного длиннее(примерно на 1000 символов).

In [298]:
forms = list(forms)
bias = pd.DataFrame(columns=['Слово','Часть речи','Одушевленность', 'Вид','Падеж','Род','Наклонение','Число',
                             'Лицо','Время','Переходность','Залог','Нормальная форма'],
                      index=[i for i in range(len(forms))])
bias["Слово"] = forms
bias["Часть речи"] = [morph.parse(i)[0].tag.POS for i in forms]
bias['Одушевленность'] = [morph.parse(i)[0].tag.animacy for i in forms]
bias['Вид'] = [morph.parse(i)[0].tag.aspect for i in forms]
bias['Падеж'] = [morph.parse(i)[0].tag.case for i in forms]
bias['Род'] = [morph.parse(i)[0].tag.gender for i in forms]
bias['Наклонение'] = [morph.parse(i)[0].tag.mood for i in forms]
bias['Число'] = [morph.parse(i)[0].tag.number for i in forms]
bias['Лицо'] = [morph.parse(i)[0].tag.person for i in forms]
bias['Время'] = [morph.parse(i)[0].tag.tense for i in forms]
bias['Переходность'] = [morph.parse(i)[0].tag.transitivity for i in forms]
bias['Залог'] = [morph.parse(i)[0].tag.voice for i in forms]
bias["Нормальная форма"] = [morph.parse(i)[0].normal_form for i in forms]
bias.to_html("martin.html")

Как и в предыдущем примере, табличку можно посмотреть отдельно, она будет в отдельном файле martin.html

Начнем вычислять статистические характеристики. Выполним сбор тех же характеристик, что и в прошлом пункте, чтобы сравнить результаты.

In [299]:
new_bias = pd.DataFrame(columns=['Слово','Часть речи','Одушевленность', 'Вид','Падеж','Род','Наклонение','Число',
                             'Лицо','Время','Переходность','Залог','Нормальная форма'],
                      index=[i for i in range(len(words))])
new_bias["Слово"] = words
new_bias["Часть речи"] = [morph.parse(i)[0].tag.POS for i in words]
new_bias['Одушевленность'] = [morph.parse(i)[0].tag.animacy for i in words]
new_bias['Вид'] = [morph.parse(i)[0].tag.aspect for i in words]
new_bias['Падеж'] = [morph.parse(i)[0].tag.case for i in words]
new_bias['Род'] = [morph.parse(i)[0].tag.gender for i in words]
new_bias['Наклонение'] = [morph.parse(i)[0].tag.mood for i in words]
new_bias['Число'] = [morph.parse(i)[0].tag.number for i in words]
new_bias['Лицо'] = [morph.parse(i)[0].tag.person for i in words]
new_bias['Время'] = [morph.parse(i)[0].tag.tense for i in words]
new_bias['Переходность'] = [morph.parse(i)[0].tag.transitivity for i in words]
new_bias['Залог'] = [morph.parse(i)[0].tag.voice for i in words]
new_bias["Нормальная форма"] = [morph.parse(i)[0].normal_form for i in words]

### 1) Длина предложения.

In [300]:
print (len(martin_text.replace('\n   ','').split('.')), 'предложений')
lens = [len(i.split(' ')) for i in martin_text.replace('\n   ','').split('.')]
c = 0
for i in lens:
    c += i
c/len(martin_text.replace('\n   ','').split('.'))

284 предложений


13.704225352112676

Средняя длина предложений в новом тексте состовляет почти 14 слов . В предыдущем тексте она была 18. 

### 2) Наиболее частый падеж.

In [301]:
new_bias["Падеж"].describe()

count     1743
unique       7
top       nomn
freq       654
Name: Падеж, dtype: object

Логично предположить, что именительный падеж снова будет самым популярным в тексте. Однако в новом отрывке его процентное содержание состовляет
практически 37 процентов, в то время как в прошлом было около 50 процентов.

### 3) Относительная частота падежей.

In [302]:

cases_table = pd.DataFrame(columns=['Частота'], index = ['Именительный падеж',
'Родительный падеж','Винительный падеж','Дательный падеж','Творительный падеж','Предложный падеж'])
freq = [len(new_bias[new_bias["Падеж"]=='nomn'])/1743, len(new_bias[new_bias["Падеж"]=='gent'])/1743,len(new_bias[new_bias["Падеж"]=='accs'])/1743,
       len(new_bias[new_bias["Падеж"]=='datv'])/1743,len(new_bias[new_bias["Падеж"]=='ablt'])/1743,len(new_bias[new_bias["Падеж"]=='loct'])/1743]
cases_table['Частота'] = freq
cases_table

Частота
Именительный падеж  0.375215
Родительный падеж   0.258749
Винительный падеж   0.145152
Дательный падеж     0.067699
Творительный падеж  0.077453
Предложный падеж    0.072863

Самый редкий падеж в этом тексте также предложный(меньше одного процента, как и в предыдущем тексте).

### 4) Процент разных частей речи.

In [303]:
speech = ['Глагол','Существительное','Прилагательное (полное)','Местоимение-существительное','Причастие (полное)','Предлог',
 'Деепричастие','Наречие','Глагол (инфинитив)','Частица','Прилагательное (краткое)','Союз','Причастие (краткое)','Числительное',
 'Предикатив' ,'Компаратив' ,'Междометие']
speech_part_table = pd.DataFrame(columns=['Процентное содержание'], index = speech)
en_speech = ['VERB', 'NOUN', 'ADJF', 'NPRO', 'PRTF', 'PREP',  'GRND',
       'ADVB', 'INFN', 'PRCL', 'ADJS', 'CONJ', 'PRTS', 'NUMR', 'PRED',
       'COMP', 'INTJ']
freq = [len(new_bias[new_bias["Часть речи"]==i])/1743*100 for i in en_speech]
speech_part_table['Процентное содержание'] = freq
speech_part_table

Процентное содержание
Глагол                                   27.710843
Существительное                          47.676420
Прилагательное (полное)                  23.235800
Местоимение-существительное              26.219162
Причастие (полное)                        2.696500
Предлог                                  20.367183
Деепричастие                              2.639128
Наречие                                  12.105565
Глагол (инфинитив)                        5.048766
Частица                                  10.786001
Прилагательное (краткое)                  2.237522
Союз                                     22.317843
Причастие (краткое)                       0.229489
Числительное                              0.172117
Предикатив                                0.573723
Компаратив                                0.458979
Междометие                                0.286862

Самая частая часть речи - существительное, совпадает с предыдущим. Но наконец-то появилось первое заметное различие, не совпадает самая редкая часть речи.
В этом тексте это числительное, а в тексте про Алису это было причастие.

### 5) Самые частотные слова и их относительная частота.

In [304]:
from collections import Counter
exclude = set(string.punctuation)
def test_set(s):
    return ''.join(ch for ch in s if ch not in exclude)
words = martin_text.split()
words = [test_set(i) for i in words]
words = [i for i in words if i != '']
cap_words = [word.upper() for word in words] #capitalizes all the words

word_counts = Counter(cap_words) 

In [305]:
for i in range(10):
    print ((word_counts.most_common()[i]), word_counts.most_common()[i][1]/len(words))

('И', 153) 0.042773273693038856
('ОН', 135) 0.03774112384679899
('В', 87) 0.024322057590159353
('НА', 70) 0.019569471624266144
('НЕ', 54) 0.015096449538719598
('ЧТО', 54) 0.015096449538719598
('ЕГО', 54) 0.015096449538719598
('ОНА', 40) 0.01118255521386637
('КАК', 38) 0.01062342745317305
('С', 35) 0.009784735812133072


Самым частым словом снова является предлог "и". Однако следующие места отличаются. Предлоги и местоимения неизменно являются лидерами по
появлению в текстах.

### 6) Самые частотные слова основных частей речи (существительные, прилагательные, наречия, глаголы).

In [306]:
nouns = new_bias[new_bias["Часть речи"]=='NOUN']
nouns.describe()

Слово Часть речи Одушевленность  Вид Падеж   Род  Наклонение Число  \
count     831        831            830  0.0   831   828         0.0   831   
unique    610          1              2  0.0     7     3         0.0     2   
top     книги       NOUN           inan  NaN  gent  femn         NaN  sing   
freq        8        831            688  NaN   252   342         NaN   683   

        Лицо  Время  Переходность  Залог Нормальная форма  
count    0.0    0.0           0.0    0.0              831  
unique   0.0    0.0           0.0    0.0              447  
top      NaN    NaN           NaN    NaN            книга  
freq     NaN    NaN           NaN    NaN               18

Неудивительно, что в романе Мартин Иден слово книга является самым частым среди всех существительных(8 раз)

In [307]:
adj = new_bias[new_bias["Часть речи"]=='ADJF']
adj.describe()

Слово Часть речи Одушевленность  Вид Падеж   Род  Наклонение Число  \
count    405        405             24  0.0   405   280         0.0   405   
unique   301          1              1  0.0     6     3         0.0     2   
top      все       ADJF           inan  NaN  nomn  femn         NaN  sing   
freq      18        405             24  NaN   153   132         NaN   280   

        Лицо  Время  Переходность  Залог Нормальная форма  
count    0.0    0.0           0.0    0.0              405  
unique   0.0    0.0           0.0    0.0              193  
top      NaN    NaN           NaN    NaN             весь  
freq     NaN    NaN           NaN    NaN               31

Слово все снова выбилось в топ(18 раз)

In [308]:
adv = new_bias[new_bias["Часть речи"]=='ADVB']
adv.describe()

Слово Часть речи  Одушевленность  Вид  Падеж  Род  Наклонение  Число  \
count     211        211             0.0  0.0    0.0  0.0         0.0    0.0   
unique    123          1             0.0  0.0    0.0  0.0         0.0    0.0   
top     вдруг       ADVB             NaN  NaN    NaN  NaN         NaN    NaN   
freq        9        211             NaN  NaN    NaN  NaN         NaN    NaN   

        Лицо  Время  Переходность  Залог Нормальная форма  
count    0.0    0.0           0.0    0.0              211  
unique   0.0    0.0           0.0    0.0              114  
top      NaN    NaN           NaN    NaN            вдруг  
freq     NaN    NaN           NaN    NaN                9

Наречие вдруг является наиболее популярным(9 раз).

In [309]:
ver = new_bias[new_bias["Часть речи"]=='VERB']
ver.describe()

Слово Часть речи  Одушевленность   Вид  Падеж   Род Наклонение Число  \
count    483        483             0.0   483    0.0   364        483   483   
unique   339          1             0.0     2    0.0     3          2     2   
top     было       VERB             NaN  perf    NaN  masc       indc  sing   
freq      14        483             NaN   254    NaN   224        480   418   

        Лицо Время Переходность  Залог Нормальная форма  
count     59   480          483    0.0              483  
unique     3     3            2    0.0              281  
top     3per  past         intr    NaN             быть  
freq      38   418          264    NaN               39

Глаголы совпадают. Даже частота(14 и 12) очень похожи.

### 7)Общее число словоупотреблений. (число слов в тексте)

In [310]:
words = martin_text.split()
exclude = set(string.punctuation)
exclude.add('- ')
def test_set(s):
    return ''.join(ch for ch in s if ch not in exclude)
words = [test_set(i) for i in words]
words = [i for i in words if i != '']
print ('Количество словоупотреблений',len(words))


Количество словоупотреблений 3577


### 8) Число различных словоформ.

In [311]:
print ('Количество словоформ',len(forms))


Количество словоформ 1774


### 9) Наиболее частотные морфологические формы глаголов (время/лицо/число).

In [312]:
new_bias["Время"].describe()

count      587
unique       3
top       past
freq       475
Name: Время, dtype: object

In [313]:
new_bias["Лицо"].describe()

count      464
unique       3
top       3per
freq       367
Name: Лицо, dtype: object

In [314]:
new_bias["Число"].describe()

count     2266
unique       2
top       sing
freq      1859
Name: Число, dtype: object

Совпадат все характеристики (прошедшее время, 3-е лицо, единственное число).

Как видно из сравнения, несмотря на то, что мы сравнивали двух разных авторов, которые писали для разных возрастов и в разное время, многие 
характеристики совпадают. Работать с данной библиотекой было очень удобно. 
Самым логичным оказалось то, что наиболее частые слова в текстах были предлогами и местоимениями, 
а самым частым падежом - именительный падеж. 
Так же интересным фактом для меня оказалось то, что в тексте Джека Лондона, в отличии от первого, самым частым является не слово "Мартин", а слово "книга". Хотя, это оказывается логичным, ведь главный герой очень увлекался книгами и мечтает стать писателем. 
Проделанная работа показалась мне довольно интересной. 